r/src/r_to_arrow.cpp

```diff
@@ -727,9 +727,33 @@ class RPrimitiveConverter<T, enable_if_t<is_timestamp_type<T>::value>>
 template <typename T>
 class RPrimitiveConverter<T, enable_if_t<is_decimal_type<T>::value>>
     : public PrimitiveConverter<T, RConverter> {
+  using ValueType = typename arrow::TypeTraits<T>::BuilderType::ValueType;
+
  public:
   Status Extend(SEXP x, int64_t size, int64_t offset = 0) override {
-    return Status::NotImplemented("Extend");
+    RETURN_NOT_OK(this->Reserve(size - offset));
+    int32_t precision = this->primitive_type_->precision();
+    int32_t scale = this->primitive_type_->scale();
+
+    auto append_value = [this, precision, scale](double value) {
+      auto converted = ValueType::FromReal(value, precision, scale);
+      RETURN_NOT_OK(converted);
+      this->primitive_builder_->UnsafeAppend(converted.ValueUnsafe());
```

Comment:
```suggestion
      ARROW_ASSIGN_OR_RAISE(ValueType converted, ValueType::FromReal(value, precision, scale));
      this->primitive_builder_->UnsafeAppend(converted);
```
You can use `ARROW_ASSIGN_OR_RAISE` here.  Also, I'm personally not a huge fan of auto but YMMV.